src/napari_threedee/annotators/label/annotator.py

```diff
@@ -0,0 +1,208 @@
+            add_mouse_callback_safe(self.labels_layer.mouse_drag_callbacks, self.draw)
+
+        # currently the plane painting must be done with n_edit_dims = 3
+        # todo modify painting so it can be done 2D in plane
+        self.labels_layer.n_edit_dims = 3
+
+    def _on_disable(self):
+        if self.labels_layer is None:
+            return
+        self.labels_layer.events.mode.disconnect(self._on_mode_change)
+
+        # switch the mouse callbacks
+        remove_mouse_callback_safe(self.labels_layer.mouse_drag_callbacks, self.draw)
+        if self.labels_layer.mode in ("paint", "erase"):
+            add_mouse_callback_safe(self.labels_layer.mouse_drag_callbacks, napari_draw)
+
+    def _mouse_event_to_nd_line_plane_intersection(self, layer, event):
+        painting_plane = self.image_layer.plane
+        intersection_3d = painting_plane.intersect_with_line(
+            line_position=layer._world_to_displayed_data(
+                position=event.position, dims_displayed=event.dims_displayed
+            ),
+            line_direction=layer._world_to_displayed_data_ray(
+                event.view_direction, event.dims_displayed
+            ),
+        )
+        intersection_nd = np.copy(event.position)
+        intersection_nd[event.dims_displayed] = intersection_3d
+        return intersection_nd
+
+    def _mouse_event_to_labels_coordinate(self, layer, event):
+        """Return the data coordinate of a Labels layer mouse event in 2D or 3D.
+
+        In 2D, this is just the event's position transformed by the layer's
+        world_to_data transform.
+
+        In 3D, a ray is cast in data coordinates, and the coordinate of the first
+        nonzero value along that ray is returned. If the ray only contains zeros,
+        None is returned.
+
+        Parameters
+        ----------
+        layer : napari.layers.Labels
+            The Labels layer.
+        event : vispy MouseEvent
+            The mouse event, containing position and view direction attributes.
+
+        Returns
+        -------
+        coordinates : array of int or None
+            The data coordinates for the mouse event.
+        """
+        ndim = len(layer._dims_displayed)
+        if ndim == 2:
+            coordinates = layer.world_to_data(event.position)
+        elif ndim == 3 and self.image_layer._depiction == VolumeDepiction.PLANE:
+            # image layer is in plane rendering, so we should paint on the plane
+            return self._mouse_event_to_nd_line_plane_intersection(layer, event)
+        else:  # 3d
+            start, end = layer.get_ray_intersections(
+                position=event.position,
+                view_direction=event.view_direction,
+                dims_displayed=layer._dims_displayed,
+                world=True,
+            )
+            if start is None and end is None:
+                return None
+            coordinates = first_nonzero_coordinate(layer.data, start, end)
+        return coordinates
```

Comment:
logic here is super clean, nice work

Reply:
Haha, I'm glad you think so, @alisterburt  - it's largely copied from your PR to napari 😆